### PR TITLE
[SPARK-42392][CORE] Add a new case of `TriggeredByExecutorDecommissionInfo` to remove unnecessary param

### PR DESCRIPTION
**Pull request by smallzhongfeng**
### What changes were proposed in this pull request?

For `TODO: add a new type like 'ExecutorDecommissionInfo' for the case where executor is decommissioned at executor first, so we don't need this extra parameter`, I add a new case class named TriggeredByExecutorDecommissionInfo to resolve it.

### Why are the changes needed?

Remove unnecessary param.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Fix uts.

### CODE REVIEW

**Comment by smallzhongfeng** [comment]
Could you help me review this? @HyukjinKwon @LuciferYang @Ngone51 Thanks a lot. :-)

**Comment by smallzhongfeng** [comment]
Gentle ping. @LuciferYang @Ngone51 @dongjoon-hyun @HyukjinKwon :-)

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala [hunk 1]
```diff
@@ -25,7 +25,19 @@ package org.apache.spark.scheduler
  *                shuffle data might be lost even if the external shuffle service is enabled.
  */
 private[spark]
-case class ExecutorDecommissionInfo(message: String, workerHost: Option[String] = None)
+abstract class DecommissionInfo
+
+private[spark]
+case class ExecutorDecommissionInfo(
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:31]: hmm... Who triggered this? And I think we should  add some comments for these two new classes

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala [hunk 2]
```diff
@@ -25,7 +25,19 @@ package org.apache.spark.scheduler
  *                shuffle data might be lost even if the external shuffle service is enabled.
  */
 private[spark]
-case class ExecutorDecommissionInfo(message: String, workerHost: Option[String] = None)
+abstract class DecommissionInfo
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:28]: If all subclasses must to be defined in this file, maybe `sealed`?

File: core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala [hunk 3]
```diff
@@ -111,21 +111,24 @@ private[spark] trait ExecutorAllocationClient {
...
-      triggeredByExecutor: Boolean = false): Boolean = {
-    val decommissionedExecutors = decommissionExecutors(
-      Array((executorId, decommissionInfo)),
-      adjustTargetNumExecutors = adjustTargetNumExecutors,
-      triggeredByExecutor = triggeredByExecutor)
+      decommissionInfo: DecommissionInfo,
+      adjustTargetNumExecutors: Boolean): Boolean = {
+    val decommissionedExecutors = decommissionInfo match {
+      case _: ExecutorDecommissionInfo =>
+        decommissionExecutors(
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala:122]: A little confused. Seems `decommissionExecutors` don't use `triggeredByExecutor` flag?

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala (hunk 1 above)
Comment (smallzhongfeng) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:31]: `Executor` will triggered this. Maybe we can see this comment https://github.com/apache/spark/blob/1c0bd1f9f813a341bbfdecb2c5ccde7fbc1bac2d/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala#L109

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:31]: Could we also rename this one to `TriggeredByXXXDecommissionInfo`?

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala (hunk 2 above)
Comment (smallzhongfeng) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:28]: OK, sounds good!

File: core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala (hunk 3 above)
Comment (smallzhongfeng) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala:122]: Actually `triggeredByExecutor` was used in `ExecutorAllocationClient` 's implementation class `CoarseGrainedSchedulerBackend` and `KubernetesClusterSchedulerBackend`. We can see this https://github.com/apache/spark/blob/9f7582c8cbeae70c31c183567bf5320d1c3210fe/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L555

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala:122]: cc @pan3793 FYI

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala (hunk 1 above)
Comment (smallzhongfeng) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:31]: Yes, how about `ExecutorTriggerExecutorDecommissionInfo` and `DriverTriggerExecutorDecommissionInfo`? Or could you give me a better opinion ?

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:31]: fine to me, it's good to be able to distinguish them clearly

File: core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala (hunk 3 above)
Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/ExecutorAllocationClient.scala:122]: OK， I see

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala [hunk 4]
```diff
@@ -17,6 +17,9 @@
 
 package org.apache.spark.scheduler
 
+private[spark]
+sealed abstract class DecommissionInfo
```
[the remark this answers is not in the document]
Comment (smallzhongfeng) [reply core/src/main/scala/org/apache/spark/scheduler/ExecutorDecommissionInfo.scala:21]: It should not be necessary. After all, the case class here only needs to inherit this class `DecommissionInfo`. If the case class need multiple inheritance, it is more appropriate to use the trait here.

**Pull request closed without merging**